second ludo review of group 5 by group 3 and 8

skjohansen opened the pull request.

Bluemace committed:
@bluemace

Added link to google doc

Bluemace committed:
@bluemace

Uppdaterad Länk till rätt dokumentation

Bluemace committed:
@bluemace @andremorad  @singelmamma

Ändrar i kod för omstrukturering

Bluemace committed:
@bluemace @andremorad @singelmamma

Moved all the classes, the database and migrations into "GameEngine" and added "CreatePlayer" Method.

singelmamma committed:
@singelmamma

AndreMorad committed:
@andremorad

reinstall

singelmamma committed:
@singelmamma

Uppdaterat så migration fungerar via GameEngine

singelmamma committed:
@Singelmamma

Added checkwinner method and some unit tests

singelmamma committed:
@Singelmamma

Added new method for player to choose tokens

singelmamma committed:
@singelmamma

Added new method for playerturn

AndreMorad committed:
@andremorad

change in createPlayer

AndreMorad committed:
@andremorad

working with createPlayer

AndreMorad committed:
@andremorad

new organization

singelmamma committed:
@singelmamma

Sorted out a lot, created new parts to the game engine and made a demo game from start to finish for players with one token

singelmamma committed:
@singelmamma

Another update on demogame

singelmamma committed:
@singelmamma @andremorad @bluemace

Bluemace committed:
Added Documentation

singelmamma committed:
@singelmamma

Added comments and deleted DieClass

singelmamma committed:
Merge branch 'master' of https://github.com/PGBSNH19/the-ludo-game-grupp-5

singelmamma committed:
@singelmamma

Updated documentation

singelmamma committed:
@singelmamma

added doc-files

simbasang commented on Source/TheLudoGameApp/Classes/GameMenu.cs:
```diff
@@ -0,0 +1,75 @@
+        {
+            Console.WriteLine("How many players? ");
+
+            int amountOfPlayers = int.Parse(Console.ReadLine());
```
tänk på att använda en try här så att programmet inte kan krascha

skjohansen replied:
Man kan med fördel använda sig av metoden TryParse

simbasang commented on Source/TheLudoGameApp/Classes/GameMenu.cs:
```diff
@@ -0,0 +1,75 @@
+        /*For the moment a gameflow prototype*/
+        public void InGame(Game game)
+        {
+            bool loop = false;
```
kanske ha ett tydligare namn?

singelmamma replied:
Jag antar att ni menar bool variabeln? Och ja galet namn, ändrar på det! Tack för input! :)

kaptenen commented on Source/TheLudoGameEngine/Classes/Game.cs:
```diff
@@ -0,0 +1,71 @@
+            Blue,
+            Green,
+            Yellow
+        }
```
Bra att ha en enum på färgerna

kaptenen commented on Source/TheLudoGameEngine/Database/MyContext.cs:
```diff
@@ -4,13 +4,14 @@
+
         protected override void OnConfiguring(DbContextOptionsBuilder dbContext)
         {
             dbContext.UseSqlServer(@"Data Source=den1.mssql8.gear.host;Initial Catalog=dbtheludogame;User id=dbtheludogame;password=Ip5ych-!9Vb1;");
```
Server password i klartext

simbasang commented on Source/TheLudoGameApp/Classes/MainMenu.cs:
```diff
@@ -0,0 +1,64 @@
+░░░╚═╝░░░╚═╝░░╚═╝╚══════╝  ╚══════╝░╚═════╝░╚═════╝░░╚════╝░░╚═════╝░╚═╝░░╚═╝╚═╝░░░░░╚═╝╚══════╝
+
+");
+        }
```
SNYGGT!

BenYtt commented on Source/TheLudoGameApp/Classes/GameMessages.cs:
```diff
@@ -0,0 +1,78 @@
+                    {
+                        Console.WriteLine($"{token.TokenColor} {token.TokenNumber}: In Goal");
+                    }
+                }
```
Här skulle man kunna refaktorera till linq för att slippa dubbla foreach-loopar.

simbasang commented on Source/TheLudoGameApp/Program.cs:
```diff
@@ -0,0 +1,17 @@
+        {
+            MainMenu.Menu();
+        }
+
```
Snyggt med bara en metod

kaptenen commented on Source/TheLudoGameEngine/Classes/Game.cs:
```diff
@@ -0,0 +1,71 @@
+
+        public bool CheckWinner(Player player)
+        {
+            if (player.Tokens.Where(p => p.InGoal == true).Count() == 4)
```
bokstaven p i lambdan känns lite missvisande då det är tokens ni refererar till? antar jag? När en spelares fyra tokens count är lika med fyra så vinner man? Kanske ha t i lambdan

singelmamma replied:
Ja precis, du/ni har helt rätt där. Det är något dumt muskelminne som sitter kvar från "Linq-lektionerna" från den allra första kursen, att man deklarerar lambdan med p som default. Det är något vi behöver jobba på att få bort. Tack för inputen!

singelmamma replied:
Och börja deklarera lambdan ordentligt*

simbasang commented on Source/TheLudoGameEngine/Classes/Token.cs:
```diff
@@ -0,0 +1,45 @@
+            {
+                if (this.GameBoardPosition >= 45)
+                {
+                    for (int y = i; y <= dieResult; y++)
+                    {
+                        i++;
+                        this.GameBoardPosition--;
+                    }
```
här förstår jag inte riktigt vad som händer

skjohansen replied:
Kanske kan det delas i fler metoder

singelmamma replied:
Förstår helt och hållet att du/ni inte gör det haha. En av våra regler som vi kör på är att man behöver slå jämt att gå in i mål och denna variant var någon form av stegräkning och animations experiment vi hade i början, att loopen skulle fungera som en tick för att flytta pjäsen steg för steg både fram och tillbaka istället för att addera hela tärningens utgång direkt på pjäsens nuvarande position. Metoden är helt omgjord nu efter dagens revidering :)

kaptenen commented on Source/TheLudoGameEngine/Classes/Token.cs:
```diff
@@ -0,0 +1,45 @@
+        public bool InNest { get; set; }
+        public bool InGoal { get; set; }
+
+        public int MoveToken(int dieResult)
+        {
+            for (int i = 1; i <= dieResult; i++)
+            {
+                if (this.GameBoardPosition >= 45)
+                {
+                    for (int y = i; y <= dieResult; y++)
+                    {
+                        i++;
+                        this.GameBoardPosition--;
+                    }
+                }
+                else
+                {
+                    this.GameBoardPosition++;
+                }
+            }
+            return this.GameBoardPosition;
+        }
+
+        public bool HasFinished()
+        {
+            this.InGoal = false;
+
+            if (this.GameBoardPosition == 45)
+            {
+                this.InGoal = true;
+            }
+
+            return this.InGoal;
+        }
+    }
+}
```
Denna var svår att förstå

singelmamma replied:
Förstår helt och hållet att du/ni inte gör det haha. En av våra regler som vi kör på är att man behöver slå jämt att gå in i mål och denna variant med MoveToken var någon form av stegräkning och animations experiment vi hade i början, att loopen skulle fungera som en tick för att flytta pjäsen steg för steg både fram och tillbaka istället för att addera hela tärningens utgång direkt på pjäsens nuvarande position. Idén är skrotad och har ingen ursäkt till att den har blivit ändrad sen senaste pushen men metoden är helt omgjord nu efter dagens revidering. Tack för input! :)

simbasang reviewed (commented):
Snyggt att ni läser in från databasen :)
känns som ni missat lite regler eller är det jag som missat dom till exempel att man inte får gå förbi sig själv? har ni med knuff??
har lite nästlade loopar/if-satser man kanske kan bryta ut?
men överlag snyggt jobbat!

BenYtt reviewed (commented):
Ni ser ju ut att kommit igång bra. Det hade kanske varit lite roligare om man ni hade använt lite fler regler. Som det är nu är egentligen enda regeln att man måste gå ut med 4 st av sina pjäser för att vinna. Det finns ju t.ex. egentligen regler som säger att man får gå ut med 2 i vissa fall eller att man inte får går förbi sin egen pjäs! 
Det är också en del nästling av for-loopar/if-satser på en helt del ställen, det hade varit sjyst om ni lyckades bryta ut det lite!

kaptenen reviewed (commented):
Har ni några fler regler än att gå ut på 1 eller 6. Mycket nästlade loopar. Bra med testerna

ChainsGBG commented on Source/TheLudoGameEngine/Classes/Token.cs:
```diff
@@ -0,0 +1,45 @@
+        {
+            for (int i = 1; i <= dieResult; i++)
+            {
+                if (this.GameBoardPosition >= 45)
```
Verkar bli ett "off by one" fel här så att man aldrig kan hamna på ruta 45.

singelmamma replied:
Tack för input, ja alltså den kedjade for loopen var nå form av animations och stegräknings experimentet. Då en av våra regler är att man behöver få ett jämt för att gå i mål med sin pjäs, annars behöver man backa. Själva metoden räknade stegen per varje runda i loopen(både fram och back) istället för att addera hela tärningens utgång, exempel pjäsen står på pos 43 och slår 5, så skulle den räkna 44, 45, 44, 43, 42. Så varje steg skulle vara nån form av animations motor, därav negationer i MoveToken metoden, men den idén är skrotad och metoden är ändrad till en mer rak uträkning istället för en nästlad loop. Tack för input :)

ChainsGBG reviewed (commented):
Bra struktur på koden.
Finns det någon anledning att använda negationer i era loopar?
Bra jobbat, kämpa på!

skjohansen commented on Source/TheLudoGameApp/Classes/GameMenu.cs:
```diff
@@ -0,0 +1,75 @@
+        {
+            var loadGame = engine.LoadPreviousGames();
+            GameMessages.PrintLoadGameList(loadGame);
+            InGame(loadGame[int.Parse(Console.ReadLine())]);
```
TryParse

skjohansen commented on Source/TheLudoGameApp/Classes/GameMenu.cs:
```diff
@@ -0,0 +1,75 @@
+            InGame(newGame);
+        }
+
+        public void LoadGameMenu()
```
Load? From where?

skjohansen commented on Source/TheLudoGameApp/Classes/MainMenu.cs:
```diff
@@ -0,0 +1,64 @@
+        {
+            GameMenu menu = new GameMenu();
+
+            bool mainMenu = true;
```
the name mainMenu does not reveal what this bool contains.
Boolans should have an affirmative phrase and it should also have a name so that the boolean value makes sense , like `isMainMenuShown`
https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-type-members

skjohansen commented on Source/TheLudoGameEngine/Classes/Engine.cs:
```diff
@@ -0,0 +1,86 @@
+{
+    public class Engine : Game
+    {
+        MyContext myContext = new MyContext();
```
This gives a very high coupling to MyContext.
Maybe you should also find a more meaningful name like LudoContext

skjohansen commented on Source/TheLudoGameEngine/Classes/Engine.cs:
```diff
@@ -0,0 +1,86 @@
+       */
+        public List<Token> TokensToMove(Player currentPlayer, int dieResult)
+        {
+            return currentPlayer.Tokens.Where(t => (t.InGoal == false && t.InNest == false) || ((dieResult == 1 || dieResult == 6) && t.InNest == true && t.InGoal == false)).ToList();
```
👍

skjohansen commented on Source/TheLudoGameEngine/Classes/Game.cs:
```diff
@@ -0,0 +1,71 @@
+            Round = 1;
+        }
+
+        public void CreatePlayers(string name, int color)
```
This create one player, so the name should be `CreatePlayer`

skjohansen commented on Source/UnitTestGameEngineMethods/UnitTest1.cs:
```diff
@@ -1,32 +1,73 @@
+            var testToken = new Token();
+            testToken.GameBoardPosition = 2;
+            testToken.GameBoardPosition = testToken.MoveToken(5);
+            Assert.IsTrue(testToken.GameBoardPosition == 7);
```
Use Equal: `Assert.Equal(7, testToken.GameBoardPosition)

skjohansen commented on Source/UnitTestGameEngineMethods/UnitTest1.cs:
```diff
@@ -1,32 +1,73 @@
+        }
         [TestMethod]
-        public void TestingDieThrow()
+        public void Token_MovmeToken_Forward_True()
```
MoveToken_FromPosition2And5Forward_Position7

skjohansen commented on Source/UnitTestGameEngineMethods/UnitTest1.cs:
```diff
@@ -1,32 +1,73 @@
+                }                
+            }
+            testGame.Finished = testGame.CheckWinner(testWinner);
+            Assert.IsTrue(testGame.Finished == true);
```
Assert.IsTrue(testGame.Finished);

skjohansen commented on Source/UnitTestGameEngineMethods/UnitTest1.cs:
```diff
@@ -1,32 +1,73 @@
+        {
+            var testGame = new Game();
+            var testWinner = new Player();
+            for(int i = 0; i<4; i++)
+            {
+                var testToken = new Token();
+                if (i > 2)
+                {
+                    testToken.InGoal = false;
+                    testWinner.Tokens.Add(testToken);
+                }
+                else
+                {
+                    testToken.InGoal = true;
+                    testWinner.Tokens.Add(testToken);
+                }                
+            }
```
Logic in the test?? 😧

skjohansen reviewed (commented):
In general fine. Watch your naming!

singelmamma committed:
@singelmamma @bluemace @andremorad

singelmamma committed:
@singelmamma

Update method and gitigonre on connectionsstring json-file

singelmamma committed:
Delete appsettings.json

singelmamma committed:
@singelmamma

singelmamma committed:
Delete MyContext.cs

singelmamma committed:
@singelmamma

singelmamma committed:
@singelmamma

Changed name on method

AndreMorad committed:
@andremorad

add xUnit test projekt

singelmamma committed:
@singelmamma

AndreMorad committed:
@andremorad

AndreMorad committed:
@andremorad

AndreMorad committed:
Merge branch 'master' of https://github.com/PGBSNH19/the-ludo-game-grupp-5

singelmamma committed:
@singelmamma

added test

AndreMorad committed:
@andremorad

AndreMorad committed:
Merge branch 'master' of https://github.com/PGBSNH19/the-ludo-game-grupp-5

singelmamma committed:
@singelmamma

Deleted files

singelmamma committed:
@singelmamma

Deleted more files and added tests

singelmamma committed:
@singelmamma

Added som more unit tests and renamned some of the old tests

singelmamma committed:
@singelmamma

Rearranged GameWinner test

AndreMorad committed:
@andremorad

added KnockOutAnotherToken xUnitTest

singelmamma committed:
@singelmamma

AndreMorad committed:
Merge branch 'master' of https://github.com/PGBSNH19/the-ludo-game-grupp-5

singelmamma committed:
@singelmamma

Added intergration testfile and moq package

Bluemace committed:
@bluemace

Unittest for InGoal.

AndreMorad committed:
@andremorad

added TokensToMove unit test

AndreMorad committed:
@andremorad

added Thread.Sleep and moved it to GameMenu

AndreMorad committed:
@andremorad

Changet my test  to pass

Bluemace committed:
@bluemace

Added Tests for UpdateTurnAndRound

singelmamma committed:
@singelmamma

Added new gamemessages and changed methods call in gameflow

singelmamma committed:
@singelmamma

"Token knocked out antoher token with the same color" bugfixed

singelmamma committed:
@singelmamma

Added more unit test

AndreMorad committed:
@andremorad

some changes

singelmamma committed:
@singelmamma

singelmamma committed:
Delete appsettings.dev.json

AndreMorad committed:
@andremorad

AndreMorad committed:
Merge branch 'master' of https://github.com/PGBSNH19/the-ludo-game-grupp-5

singelmamma committed:
@singelmamma

singelmamma committed:
Merge branch 'master' of https://github.com/PGBSNH19/the-ludo-game-grupp-5

singelmamma committed:
@singelmamma

singelmamma committed:
@singelmamma @andremorad  @bluemace

Uppdaterade spelflödet och testerna

singelmamma committed:
@singelmamma @andremorad

Rearranged the tests to different testfiles and added som more tests

AndreMorad committed:
@andremorad

added new test

AndreMorad committed:
@andremorad

added colors to console

singelmamma committed:
@singelmamma

singelmamma committed:
@singelmamma

singelmamma committed:
@singelmamma

singelmamma committed:
@singelmamma @andremorad @bluemace

Reviderade koden och tog bort onödiga metoden, samt ny namngivning på vissa metoder och varibler

singelmamma committed:
@singelmamma

singelmamma committed:
@singelmamma

singelmamma committed:
@singelmamma

singelmamma committed:
@singelmamma @andremorad  @bluemace

singelmamma committed:
@singelmamma

singelmamma committed:
@singelmamma
